pep-0624.rst

```diff
@@ -0,0 +1,284 @@
+``const char *errors`` or ``int byteorder`` parameters.
+
+We can rename some private APIs and make them public to cover missing
+APIs and parameters.
+
+============================= ================================
+ Rename to                     Rename from
+============================= ================================
+``PyUnicode_EncodeASCII()``    ``_PyUnicode_AsASCIIString()``
+``PyUnicode_EncodeLatin1()``   ``_PyUnicode_AsLatin1String()``
+``PyUnicode_EncodeUTF7()``     ``_PyUnicode_EncodeUTF7()``
+``PyUnicode_EncodeUTF8()``     ``_PyUnicode_AsUTF8String()``
+``PyUnicode_EncodeUTF16()``    ``_PyUnicode_EncodeUTF16()``
+``PyUnicode_EncodeUTF32()``    ``_PyUnicode_EncodeUTF32()``
+============================= ================================
+
+Pros:
+
+* We have more consistent API set.
+
+Cons:
+
+* We have more public APIs to maintain.
+* Existing public APIs are enough for most use cases, and
+  ``PyUnicode_AsEncodedString()`` can be used in other cases.
+
+
+Replace ``Py_UNICODE*`` with ``Py_UCS4*``
+-----------------------------------------
+
+We can replace ``Py_UNICODE`` (typedef of ``wchar_t``) with
+``Py_UCS4``. Since builtin codecs support UCS-4, we don't need to
+convert ``Py_UCS4*`` string to Unicode object.
+
+
+Pros:
+
+* We have more consistent API set.
+* User can encode UCS-4 string in C without creating Unicode object.
+
+Cons:
+
+* We have more public APIs to maintain.
+* Applications which uses UTF-8 or UTF-32 can not use these APIs
+  anyway.
+* Other Python implementations may not have builtin codec for UCS-4.
+* If we change the Unicode internal representation to UTF-8, we need
+  to keep UCS-4 support only for these APIs.
+
+
+Replace ``Py_UNICODE*`` with ``wchar_t*``
+-----------------------------------------
+
+We can replace ``Py_UNICODE`` to ``wchar_t``.
+
+Pros:
+
+* We have more consistent API set.
+* Backward compatible.
+
+Cons:
+
+* We have more public APIs to maintain.
+* They are inefficient on platforms ``wchar_t*`` is UTF-16. It is
+  because built-in codecs supports only UCS-1, UCS-2, and UCS-4
+  input.
+
+
+Rejected ideas
+==============
+
+Using runtime warning
+---------------------
+
+These APIs doesn't release GIL for now. Emitting a warning from
+such APIs is not safe. See this example.
+
+.. code-block::
+
+   PyObject *u = PyList_GET_ITEM(list, i);  // u is borrowed reference.
+   PyObject *b = PyUnicode_EncodeUTF8(PyUnicode_AS_UNICODE(u),
+           PyUnicode_GET_SIZE(u), NULL);
+   // Assumes u is still living reference.
+   PyObject *t = PyTuple_Pack(2, u, b);
+   Py_DECREF(b);
+   return t;
+
+If we emit Python warning from ``PyUnicode_EncodeUTF8()``, warning
+filters and other threads may change the ``list`` and ``u`` can be
+a dangling reference after ``PyUnicode_EncodeUTF8()`` returned.
+
+Additionally, since we are not changing behavior but removing C APIs,
+runtime ``DeprecationWarning`` might not helpful for Python
+developers. We should warn to extension developers instead.
+
```

Comment:
Would you mind to add a section with links to previous discussions/issues? Something like:

```
Discussions
===

* `Plan to remove Py_UNICODE APis except PEP 623
  <https://mail.python.org/archives/list/python-dev@python.org/thread/S7KW2U6IGXZFBMGS6WSJB26NZIBW4OLE/#S7KW2U6IGXZFBMGS6WSJB26NZIBW4OLE>`_
* `bpo-41123: Remove Py_UNICODE APIs except PEP 623: <https://bugs.python.org/issue41123>`_
```

Are you aware of other past discussions?

Since PEP 623 is related but different, it may also help to mention it somewhere in this PEP, and maybe explain that why they are independent.